### ISSUE TITLE
The Need for Speed ​​SE (1996) PC - Stuttering

### ISSUE THREAD

**Issue by jericho22** [narzoul/DDrawCompat#334]
When I try latest version, game crashes instantly (it won't work with any other wrappers anyway).
The game only runs on dxwnd without any issues.
It also works natively, but there is a color issue.
Is there any support possible in the future?

![NFS1_new_look](https://github.com/narzoul/DDrawCompat/assets/98463237/8408028a-c542-4aaa-90e9-ddfe1986b385)

**Comment by narzoul** [comment]
> When I try latest version, game crashes instantly (it won't work with any other wrappers anyway).

Does that mean it worked with an older version? If so, which one?

> It also works natively, but there is a color issue.

Unfortunately it doesn't work for me, and the crashes don't seem to be related to DirectDraw. Did you install some patches or compatibility fixes to make it work? What version of Windows are you using?

> Is there any support possible in the future?

Only if it's actually a bug on DDrawCompat's side, which I find rather unlikely in this case. I don't have time to patch non-graphics related crashes or bugs in games.

That said, since you can apparently get further than I can natively, if you enable LogLevel=debug and CrashDump=full in the configuration, and sent both the log file and the crash dump to me, I can make sure it's at least not a crash caused by DDrawCompat.

**Comment by jericho22** [comment]
> Does that mean it worked with an older version? If so, which one?
I think doesn't work on any version.

Crashes must be ddraw related (unless your files are missing or you have not set cd letter correctly). I even checked via debug tool. An address (must be related to window) does not take any value and this causes it to crash. Gho (dxwnd developer) said that this was an issue with starting with a window that was too large, like 100000x100000. But I think there's more to it than that. Of course I'm talking about dxwnd.
I'm more interested in game modding. and I have no knowledge about renders/graphics. But I will try to help you as much as I can.

> What version of Windows are you using?

Windows 10 pro 64 bit (22H2)

Of course, some compatibility is needed to run game natively.
- GlobalMemoryStatusLie
- MapMemoryB0000
- SingleProcAffinity
- CorrectCreateSurface

Already some fixes are included in exe. and added Nocd patch by me.
- Cpu affinity
- Initial window resolution fix (height)
- Initial window resolution fix (width)
- Numlock fix
- Handle exceptions fix
- 8-bit color error fix
- NoCD patch
You can use this exe that I attached.
[NFS.zip](https://github.com/user-attachments/files/16072529/NFS.zip)

Finally, you may need to run repeatedly for it to work.


> Only if it's actually a bug on DDrawCompat's side, which I find rather unlikely in this case. I don't have time to patch non-graphics related crashes or bugs in games.

NP, I believe it will work at least just fixing ddraw part. I don't expect you to be able to fix anything non graphics anyway. If I had known, I would have fixed completely myself. But as I said, I'm not knowledgeable about graphics.

Here is dump you asked for from me (only patched exe was used. no compatibility):
https://www.sendspace.com/file/u9ay1c

**Comment by jericho22** [comment]
I forgot to mention, in version 0.5.2, the error message given by the game is:
![Screenshot_11](https://github.com/narzoul/DDrawCompat/assets/98463237/5f18a04f-ac68-4b3a-b1a4-46c4f403ca3f)

**Comment by BEENNath58** [comment]
> CorrectCreateSurface

If I am not wrong, this is the setting that causes the colour issue. As to what I remember, I have successfully tested the game without this shim

**Comment by BEENNath58** [comment]
@jericho22 The initgraphics error came from a bad synchornisation of the game threads. If you followed the DxWnd discussion, nervoushammer posted another executable that mitigated this issue.

With the 4 shims, and that executable, it works with DDrawCompat. However the game is equally slow as using DxWnd primary surface, the game is faster in D3D9 colour emulation for some reason

You can incorporate thosechanges in your executable

![Image](https://github.com/user-attachments/assets/cc659fd6-d407-410a-a671-5a511b4cf676)

**Comment by jericho22** [comment]
> [@jericho22](https://github.com/jericho22) The initgraphics error came from a bad synchornisation of the game threads. If you followed the DxWnd discussion, nervoushammer posted another executable that mitigated this issue.
> 
> With the 4 shims, and that executable, it works with DDrawCompat. However the game is equally slow as using DxWnd primary surface, the game is faster in D3D9 colour emulation for some reason
> 
> You can incorporate thosechanges in your executable
> 
> ![Image](https://github.com/user-attachments/assets/cc659fd6-d407-410a-a671-5a511b4cf676)

Thank you As you said. the game now works via ddrawcompat. To fix extreme lag/stuttering, set **ThreadPriorityBoost** to **ON**. This will fix. But not 100%. Will be sudden drops sometimes. Maybe @narzoul can provide a suggestion or fix for this?

**Comment by jericho22** [comment]
Forgot to mention If you set game's priority to **low** in the Task manager, should be fine without a single stutter while racing. I wrote an ASI for that, but it didn't work.

**Comment by BEENNath58** [comment]
@jericho22 Using Low in Task Manager makes the game a little more laggier instead. 

ThreadPriorityBoost made the GUI and video a little faster though the difference is negligible

**Comment by jericho22** [comment]
@BEENNath58 But how? ThreadPriorityBoost made the game run as good as in dxwnd for me (before that the game was running like 1 fps). But with the same drops as on dxwnd. I remembered that I fixed it with low priority. and now there is no problem when I am in the race. Would be nice if there was an option to set it without starting the game tho. Setting later makes the buttons in the menus not work or makes the game unstable.

One problem is that my desktop resolution is forced to 640x480 when starting the game. Despite forcing it to 2560x1440 from ini it didn't work. no such problem in subsequent nfs games.

**Comment by BEENNath58** [comment]
@jericho22 No it never ran with 1FPS for me. I ran the game at 640x480 as it was meant to be
Lol I can't find the setting that gave a super smooth experience for the game.

Do you mean you can set the game to 2560x1440? In my case, the default value (Desktop) resolution works and the game scales to my desktop

**Comment by narzoul** [comment]
PresentDelay=off might make it a little smoother. Other than that, not much can be done except using CpuAffinity=all, which does make the game prone to freeze, but I managed to get into a race one and it made the present rate pretty consistent at average 62-63 PFS. That seems to be the target of some internal FPS limiter (16 ms frame time?).

**Comment by jericho22** [comment]
no, PresentDelay=off did opposite and slowed down the game. I think it's best to keep it between 10-20 values. 
CpuAffinity=all already directly makes the game the smoothest. I've seen present rate current up to 71 and max up to 514. current is 33-34 when used on single core. but very wavy. as i said before low priority just makes the game smoother for me. in CpuAffinity=1.

> That seems to be the target of some internal FPS limiter (16 ms frame time?).

Yeah, old nfs games usually have hardcoded fps caps. For example nfs2, nfs3 and nfs4 are capped to 64. But if you pause the game, there is no limit. Less fps you have from 64, more lag/input lag you have.

Anyway, these were settings I used. and I'm using v0.6.

[DDrawCompat-NFS.zip](https://github.com/user-attachments/files/19735343/DDrawCompat-NFS.zip)

I noticed that in log I saw "Failed to apply custom display mode: 2560x1440@75". Although it should be 73, not 75. So could this be the reason why it is still 640x480?

**Comment by narzoul** [comment]
Why should it be 73? That's a very odd value. Anyway, the error message means that this particular resolution and refresh combination is not reported as supported by your display. This shouldn't normally happen. I can't tell much more without seeing the logs, but you should see what's reported as supported in a group of EnumDisplaySettingsExA logs.

For me, the game doesn't specify a refresh rate (it uses the older IDirectDraw::SetDisplayMode method, which doesn't let you choose a refresh rate), and by default the ddraw runtime prefers to choose your current desktop refresh rate in this case. It only uses 75 and 60 as fallbacks if the current desktop refresh rate is not supported at the target resolution, I think. Do you have a custom 73 refresh rate applied to your current desktop resolution, but not available for 640x480 perhaps?

**Comment by jericho22** [comment]
yes, I did via nvidia control panel. Normally my monitor supports 75hz at 1440x900 but since I want to use at 1440p, DSR only gives me 60hz for 1080p and above. As a result, I decided to create a custom resolution. Unfortunately, beyond 73hz my monitor just gives black screen. No difference for me 73-75, but there is a noticeable difference 60-73. and it's worth. Anyway if you want to see log here you go:

[DDrawCompat-NFS.log](https://github.com/user-attachments/files/19779393/DDrawCompat-NFS.log)

**Comment by jericho22** [comment]
nvm, for some reason I changed a few settings and now it works. But before that I tried all settings and it didn't work.

I realized that some settings don't work for this game. For example, when I turn antialiasing off and 8x, there is no difference. Likewise, when I changed many settings like resolution scale, AlternatePixelCenter, FontAntialiasing, texturefilter etc... I didn't see any difference. Maybe because uses software rendering?

lastly a trivial detail/issue; Scale of my desktop is decreasing while the game is running. Is this normal? It goes back to normal when I close the game. Although sometimes I do alt+tab and text looks small, which is annoying. i've had this happen in other games tho.

**Comment by narzoul** [comment]
I checked your log. You were forcing DisplayResolution=desktop with DisplayRefreshRate=60, which is why you got that error message ("Failed to apply custom display mode: 2560x1440@60"). It's expected if you try to force unsupported display modes. In this case it will fall back to using the display mode that the application wants to use. Based on the reported display modes, 60 Hz is not supported at 2560x1440. But since your desktop already seems to be using 2560x1440 at 73Hz, you should just use DisplayRefreshRate=desktop to keep using the same in the game too.

> For example, when I turn antialiasing off and 8x, there is no difference. Likewise, when I changed many settings like resolution scale, AlternatePixelCenter, FontAntialiasing, texturefilter etc... I didn't see any difference. Maybe because uses software rendering?

Yes, none of those settings affect software rendering, except FontAntialiasing, which only affects GDI text rendering (probably not used by this game).

> lastly a trivial detail/issue; Scale of my desktop is decreasing while the game is running. Is this normal? It goes back to normal when I close the game. Although sometimes I do alt+tab and text looks small, which is annoying. i've had this happen in other games tho.

This shouldn't happen with default settings. Only if you override DisplayResolution, or set an invalid combination of DisplayResolution+DisplayRefreshRate, like above.

You have a lot of settings unnecessarily specified, despite using the same value as the default. I don't recommend doing this. It also makes it difficult for me (and you) to see which settings you changed from the defaults. Better to just comment out not used settings if you want to have it in the config file anyway. I keep an up to date copy of the default values (all commented out) here:
https://github.com/narzoul/DDrawCompat/blob/master/Tools/DDrawCompat.ini

I've seen multiple people doing this and making a mess of their configs. I guess I might as well just update the logging to show some special marker if you modify some settings, but you've set them to the default value anyway. Currently it just shows the last config file that successfully "changed" the setting, even if it's the same value as before.

**Comment by jericho22** [comment]
> I checked your log. You were forcing DisplayResolution=desktop with DisplayRefreshRate=60, which is why you got that error message ("Failed to apply custom display mode: 2560x1440@60"). It's expected if you try to force unsupported display modes. In this case it will fall back to using the display mode that the application wants to use. Based on the reported display modes, 60 Hz is not supported at 2560x1440. But since your desktop already seems to be using 2560x1440 at 73Hz, you should just use DisplayRefreshRate=desktop to keep using the same in the game too.

No, not really. It was already set to desktop before this. and problem was still there. so I was trying some of your suggestions and forgot to set it to desktop again. Literally when I tried similar settings in nfs3 it was working.
I don't know what fixed this, but whatever. It's fine now.

> This shouldn't happen with default settings. Only if you override DisplayResolution, or set an invalid combination of DisplayResolution+DisplayRefreshRate, like above.

I use it at 175%, but when game launches it is set to 125%. dgv2 or dxwnd have no such problem. 
Also disabled settings that don't work for this game by adding #. Still same. I'll ask if anyone else has this problem.

I'm just trying settings, I mean not doing to mess up, I'm applying them after reading what they do in wiki, because game still doesn't work well in terms of fps.
I checked at dxwnd again and found that it works better than ddc (fps wise). Doesn't matter to me, but I've noticed that people prefer ddc more than dxwnd (at least in our discord server). 

Especially when using noactivateapp(0), I have never seen the game crash even if there is sound in the background. dxwnd can crash sometimes (this has happened many times). Graphically at least in ddc there is no problem (in coastal I saw green dots on the rocks once but I think it was a one time thing).

Would be nice if we could do something about fps. That's all I want rn. I would like to debug the game, but I have more important stuffs to do. Even if I have time, I've never patched rendering and performance side before, so not sure. or at least help someone who has already found something related to it. But other than that, the game works fine, even though I've edited title accordingly, we can close the topic if you want.

**Comment by narzoul** [comment]
> I use it at 175%, but when game launches it is set to 125%. dgv2 or dxwnd have no such problem.

This shouldn't happen unless the display resolution gets changed. So is that still happening? Can you share new debug logs with this problem?
There is no (documented) API in Windows to change the DPI scaling factor as far as I know. There are some undocumented ways described on StackOverflow, but I haven't implemented any of them:
https://stackoverflow.com/questions/35233182/how-can-i-change-windows-10-display-scaling-programmatically-using-c-sharp

But if the actual display resolution changes, then Windows tends to limit the scaling factor automatically, and lower resolutions have lower limits. This is the only way I know of that the actual DPI scale could change with DDrawCompat. But this shouldn't be happening with the default settings, so I would have to see your debug logs to see why your display resolution is still changing despite using the default display related settings, which should keep your desktop resolution.

> I checked at dxwnd again and found that it works better than ddc (fps wise).

Can you share both your DxWnd and DDrawCompat configuration so I can reproduce this? Also the version of DxWnd you are using.

> Would be nice if we could do something about fps. That's all I want rn. I would like to debug the game, but I have more important stuffs to do. Even if I have time, I've never patched rendering and performance side before, so not sure. or at least help someone who has already found something related to it.

From what I've seen from debug logs, the game is entirely rendered via the CPU. It only uses DirectDraw to ask for a surface (memory buffer) to draw pixels on, and then it draws everything with its own proprietary algorithms that are baked into the executable (or one of its DLLs). So I don't think there is anything any wrapper can do about it. Basically the game's software rendering engine would need to be reverse engineered and modded/patched directly at the binary level.

But since CpuAffinity=all solves the performance issues, I don't think it's the renderer itself that's too slow anyway. It's most likely thread scheduling issues and too much unnecessary busy waiting. But these are also things that can only be patched in the game itself.

I don't exclude the possibility that there's some issue with DDrawCompat that's slowing down the game for you. I hope I can replicate the problem with your DxWnd and DDrawCompat config so I can try to check why DxWnd would run the game any faster, which doesn't really make sense to me otherwise. But DxWnd has tons of hooks, maybe it's doing something related to audio or input handling or something else that improves performance, which DDrawCompat doesn't do.

**Comment by narzoul** [comment]
Closing due to inactivity and because the software renderer performance can't be improved by a wrapper anyway.
